From 196d7bdd32a72d44e335f84d204f3fa2c5f486c6 Mon Sep 17 00:00:00 2001
From: Yong Zhao <Yong.Zhao@amd.com>
Date: Fri, 28 Jul 2017 17:58:35 -0400
Subject: [PATCH 5642/5855] drm/amdkfd: Fix a bug that process cleanup is not
 done properly

When destroying a queue fails in nocpsch, it may leave other queues
not destroyed and process not unregistered. This commit fixes that.

Change-Id: I9dcf7cb71242aa02e5b72c0381fec19a85f3dfe4
Signed-off-by: Yong Zhao <Yong.Zhao@amd.com>
---
 drivers/gpu/drm/amd/amdkfd/kfd_device_queue_manager.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdkfd/kfd_device_queue_manager.c b/drivers/gpu/drm/amd/amdkfd/kfd_device_queue_manager.c
index 1a62b09..46f93ec 100644
--- a/drivers/gpu/drm/amd/amdkfd/kfd_device_queue_manager.c
+++ b/drivers/gpu/drm/amd/amdkfd/kfd_device_queue_manager.c
@@ -1434,9 +1434,11 @@ static int process_termination_nocpsch(struct device_queue_manager *dqm,
 
 	/* Clear all user mode queues */
 	list_for_each_entry_safe(q, next, &qpd->queues_list, list) {
-		retval = destroy_queue_nocpsch_locked(dqm, qpd, q);
-		if (retval)
-			goto out;
+		int ret;
+
+		ret = destroy_queue_nocpsch_locked(dqm, qpd, q);
+		if (ret)
+			retval = ret;
 	}
 
 	/* Unregister process */
@@ -1449,7 +1451,6 @@ static int process_termination_nocpsch(struct device_queue_manager *dqm,
 		}
 	}
 
-out:
 	mutex_unlock(&dqm->lock);
 	return retval;
 }
-- 
2.7.4

